File: handwritten/spanner/issue/go/raw_grpc_benchmark.go
package main

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"log"
	"math"
	mathrand "math/rand"
	"os"
	"sort"
	"strconv"
	"sync"
	"time"

	gapic "cloud.google.com/go/spanner/apiv1"
	"google.golang.org/api/option"
	spannerpb "google.golang.org/genproto/googleapis/spanner/v1"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	structpb "google.golang.org/protobuf/types/known/structpb"
)

func main() {
	ctx := context.Background()

	// Read benchmark configuration
	projectId := getEnv("DB_PROJECT_ID", "emulator")
	instanceId := getEnv("DB_INSTANCE", "device-tracking")
	databaseId := getEnv("DB_DATABASE", "device-tracking")
	tableName := getEnv("TABLE_NAME", "tracking.DeviceRecentActivityLog")
	readCount := getEnvInt("READ_COUNT", 1000)
	concurrency := getEnvInt("CONCURRENCY", 10)

	// Write benchmark configuration
	sampleSize := getEnvInt("SAMPLE_SIZE", 10000)
	insertCount := getEnvInt("INSERT_COUNT", 1000)
	insertConcurrency := getEnvInt("INSERT_CONCURRENCY", 110)
	batchCount := getEnvInt("BATCH_COUNT", 1)
	duplicateInsert := getEnv("DUPLICATE_INSERT", "false") == "true"

	databasePath := fmt.Sprintf("projects/%s/instances/%s/databases/%s", projectId, instanceId, databaseId)

	fmt.Printf("Target Database: %s\n", databasePath)
	fmt.Printf("Target Table: %s\n", tableName)
	fmt.Printf("Read Config: READ_COUNT=%d, CONCURRENCY=%d\n", readCount, concurrency)
	fmt.Printf("Write Config: SAMPLE_SIZE=%d, INSERT_COUNT=%d, INSERT_CONCURRENCY=%d, BATCH_COUNT=%d, DUPLICATE_INSERT=%t\n", sampleSize, insertCount, insertConcurrency, batchCount, duplicateInsert)
	fmt.Println("gRPC Connection Pool Size: 1")

	// Initialize the generated client with a single connection to match Node.js default behavior
	client, err := gapic.NewClient(ctx, option.WithGRPCConnectionPool(4))
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}
	defer client.Close()

	fmt.Println("\nCreating multiplexed session via raw gRPC...")
	session, err := client.CreateSession(ctx, &spannerpb.CreateSessionRequest{
		Database: databasePath,
		Session: &spannerpb.Session{
			Multiplexed: true,
		},
	})
	if err != nil {
		log.Fatalf("Failed to create session: %v", err)
	}
	fmt.Printf("Multiplexed Session created: %s\n", session.Name)

	defer func() {
		fmt.Println("\nBenchmark finished. Multiplexed session is left to be managed by server.")
	}()

	// ==========================================
	// 1. READ BENCHMARK
	// ==========================================
	fmt.Printf("\nStarting READ benchmark with concurrency %d...\n", concurrency)

	readDurations := make([]time.Duration, readCount)
	readSem := make(chan struct{}, concurrency)
	var readWg sync.WaitGroup

	readOverallStart := time.Now()

	for i := 0; i < readCount; i++ {
		readWg.Add(1)
		readSem <- struct{}{} // Acquire

		go func(index int) {
			defer readWg.Done()
			defer func() { <-readSem }() // Release

			randomId := mathrand.Intn(1000000) + 1

			req := &spannerpb.ExecuteSqlRequest{
				Session: session.Name,
				Sql:     fmt.Sprintf("SELECT * FROM %s WHERE deviceRecentActivityLogId = @id", tableName),
				Params: &structpb.Struct{
					Fields: map[string]*structpb.Value{
						"id": {
							Kind: &structpb.Value_StringValue{
								StringValue: fmt.Sprintf("%d", randomId),
							},
						},
					},
				},
				ParamTypes: map[string]*spannerpb.Type{
					"id": {
						Code: spannerpb.TypeCode_STRING,
					},
				},
			}

			startTime := time.Now()
			_, err := client.ExecuteSql(ctx, req)
			duration := time.Since(startTime)

			if err != nil {
				log.Printf("Read failed at index %d: %v", index, err)
			} else {
				readDurations[index] = duration
			}
		}(i)
	}

	readWg.Wait()
	readOverallDuration := time.Since(readOverallStart)
	readOpsPerSecond := float64(readCount) / readOverallDuration.Seconds()

	var readSum time.Duration
	validReadDurations := make([]time.Duration, 0, readCount)
	for _, d := range readDurations {
		if d > 0 {
			readSum += d
			validReadDurations = append(validReadDurations, d)
		}
	}

	var readAvg time.Duration
	if len(validReadDurations) > 0 {
		readAvg = readSum / time.Duration(len(validReadDurations))
	}

	sort.Slice(validReadDurations, func(i, j int) bool {
		return validReadDurations[i] < validReadDurations[j]
	})

	rp50 := percentile(validReadDurations, 0.5)
	rp90 := percentile(validReadDurations, 0.9)
	rp95 := percentile(validReadDurations, 0.95)
	rp99 := percentile(validReadDurations, 0.99)

	fmt.Println("\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
	fmt.Println("  Raw gRPC Benchmark Summary (Go - READ)")
	fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
	fmt.Printf("  Total Reads:        %d\n", readCount)
	fmt.Printf("  Concurrency:        %d\n", concurrency)
	fmt.Printf("  Total Time:         %v\n", readOverallDuration)
	fmt.Printf("  Reads/Second:       %.2f\n", readOpsPerSecond)
	fmt.Printf("  Avg Latency:        %v\n", readAvg)
	fmt.Printf("  P50 Latency:        %v\n", rp50)
	fmt.Printf("  P90 Latency:        %v\n", rp90)
	fmt.Printf("  P95 Latency:        %v\n", rp95)
	fmt.Printf("  P99 Latency:        %v\n", rp99)
	fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n")

	// ==========================================
	// 2. SLEEP 5 SECONDS
	// ==========================================
	fmt.Println("Sleeping for 5 seconds before starting writes...")
	time.Sleep(5 * time.Second)

	// ==========================================
	// 3. WRITE BENCHMARK
	// ==========================================
	fmt.Printf("\nStarting WRITE benchmark with concurrency %d...\n", insertConcurrency)

	totalRecords := insertCount * batchCount
	totalInserts := totalRecords
	if duplicateInsert {
		totalInserts = totalRecords * 2
	}

	// Pre-generate batches
	type Record struct {
		deviceRecentActivityLogId string
		deviceRecordId            string
		deviceDetailsId           string
		httpRequestDetailsId      string
		ipAddress                 string
		institutionId             string
		userId                    *string
		username                  *string
		xRequestId                string
		httpRequestLocationId     *string
		latency                   *int64
		sessionId                 string
		createdAt                 time.Time
	}

	batches := make([][]*Record, insertCount)
	for i := 0; i < insertCount; i++ {
		batch := make([]*Record, batchCount)
		for j := 0; j < batchCount; j++ {
			idx := i*batchCount + j
			deviceRecordId := sha256Hash(fmt.Sprintf("device-%d", idx%sampleSize))
			deviceDetailsId := sha256Hash(fmt.Sprintf("detail-%d", idx%sampleSize))
			httpRequestDetailsId := sha256Hash(fmt.Sprintf("request-%d", idx%sampleSize))
			httpRequestLocationIdStr := sha256Hash(fmt.Sprintf("location-%d", idx%sampleSize))
			var httpRequestLocationId *string = &httpRequestLocationIdStr

			xRequestId := generateUUID()
			sessionId := generateUUID()
			createdAt := time.Now().Add(time.Duration(idx) * time.Second)

			idInput := fmt.Sprintf("benchmark-%d%s%s%d%s", idx, deviceDetailsId, httpRequestDetailsId, createdAt.UnixMilli(), sessionId)
			deviceRecentActivityLogId := sha256Hash(idInput)

			var userId *string
			if mathrand.Float64() > 0.5 {
				uid := "user-123"
				userId = &uid
			}

			var latency *int64
			if mathrand.Float64() > 0.5 {
				lat := int64(mathrand.Intn(100))
				latency = &lat
			}

			batch[j] = &Record{
				deviceRecentActivityLogId: deviceRecentActivityLogId,
				deviceRecordId:            deviceRecordId,
				deviceDetailsId:           deviceDetailsId,
				httpRequestDetailsId:      httpRequestDetailsId,
				ipAddress:                 base64.StdEncoding.EncodeToString([]byte{192, 168, 1, 100}),
				institutionId:             "benchmark-institution",
				userId:                    userId,
				username:                  nil,
				xRequestId:                xRequestId,
				httpRequestLocationId:     httpRequestLocationId,
				latency:                   latency,
				sessionId:                 sessionId,
				createdAt:                 createdAt,
			}
		}
		batches[i] = batch
	}

	writeDurations := make([]float64, insertCount)
	var alreadyExistsCount int64
	var writeMu sync.Mutex

	writeSem := make(chan struct{}, insertConcurrency)
	var writeWg sync.WaitGroup
	writeOverallStart := time.Now()

	for i, batch := range batches {
		writeWg.Add(1)
		writeSem <- struct{}{}

		go func(index int, records []*Record) {
			defer writeWg.Done()
			defer func() { <-writeSem }()

			start := time.Now()

			mutations := make([]*spannerpb.Mutation, len(records))
			for j, rec := range records {
				values := []*structpb.Value{
					protoValue(rec.deviceRecentActivityLogId),
					protoValue(rec.deviceRecordId),
					protoValue(rec.deviceDetailsId),
					protoValue(rec.httpRequestDetailsId),
					protoValue(rec.ipAddress),
					protoValue(rec.institutionId),
					protoValue(rec.userId),
					protoValue(rec.username),
					protoValue(rec.xRequestId),
					protoValue(rec.httpRequestLocationId),
					protoValue(rec.latency),
					protoValue(rec.sessionId),
					protoValue(rec.createdAt.UTC().Format(time.RFC3339Nano)),
				}

				mutations[j] = &spannerpb.Mutation{
					Operation: &spannerpb.Mutation_Insert{
						Insert: &spannerpb.Mutation_Write{
							Table: tableName,
							Columns: []string{
								"deviceRecentActivityLogId",
								"deviceRecordId",
								"deviceDetailsId",
								"httpRequestDetailsId",
								"ipAddress",
								"institutionId",
								"userId",
								"username",
								"xRequestId",
								"httpRequestLocationId",
								"latency",
								"sessionId",
								"createdAt",
							},
							Values: []*structpb.ListValue{
								{Values: values},
							},
						},
					},
				}
			}

			commitReq := &spannerpb.CommitRequest{
				Session: session.Name,
				Transaction: &spannerpb.CommitRequest_SingleUseTransaction{
					SingleUseTransaction: &spannerpb.TransactionOptions{
						Mode: &spannerpb.TransactionOptions_ReadWrite_{
							ReadWrite: &spannerpb.TransactionOptions_ReadWrite{},
						},
					},
				},
				Mutations: mutations,
			}

			_, err := client.Commit(ctx, commitReq)
			if err != nil {
				if st, ok := status.FromError(err); ok && st.Code() == codes.AlreadyExists {
					writeMu.Lock()
					alreadyExistsCount++
					writeMu.Unlock()
				} else {
					log.Printf("Commit failed at batch %d: %v", index, err)
				}
			}

			if duplicateInsert {
				_, err = client.Commit(ctx, commitReq)
				if err != nil {
					if st, ok := status.FromError(err); ok && st.Code() == codes.AlreadyExists {
						writeMu.Lock()
						alreadyExistsCount++
						writeMu.Unlock()
					} else {
						log.Printf("Duplicate commit failed at batch %d: %v", index, err)
					}
				}
			}

			duration := time.Since(start).Milliseconds()
			writeMu.Lock()
			writeDurations[index] = float64(duration)
			writeMu.Unlock()

			if (index+1)%100 == 0 || (index+1) == insertCount {
				writeMu.Lock()
				windowStart := index - 99
				if windowStart < 0 {
					windowStart = 0
				}
				var sum float64
				count := 0
				for k := windowStart; k <= index; k++ {
					sum += writeDurations[k]
					count++
				}
				avg := sum / float64(count)
				writeMu.Unlock()
				fmt.Printf("  [%d/%d] Last batches avg: %.2fms (%d records/batch)\n", index+1, insertCount, avg, batchCount)
			}
		}(i, batch)
	}

	writeWg.Wait()
	writeOverallDuration := time.Since(writeOverallStart).Milliseconds()
	writeInsertsPerSecond := float64(totalInserts) / (float64(writeOverallDuration) / 1000.0)

	var writeSum float64
	writeMin := writeDurations[0]
	writeMax := writeDurations[0]
	for _, d := range writeDurations {
		writeSum += d
		if d < writeMin {
			writeMin = d
		}
		if d > writeMax {
			writeMax = d
		}
	}
	writeAvg := writeSum / float64(len(writeDurations))

	sortedWrite := make([]float64, len(writeDurations))
	copy(sortedWrite, writeDurations)
	sort.Float64s(sortedWrite)

	wp50 := sortedWrite[int(float64(len(sortedWrite))*0.5)]
	wp90 := sortedWrite[int(float64(len(sortedWrite))*0.9)]
	wp95 := sortedWrite[int(float64(len(sortedWrite))*0.95)]
	wp99 := sortedWrite[int(math.Min(float64(len(sortedWrite))*0.99, float64(len(sortedWrite)-1)))]

	fmt.Println("\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
	fmt.Println("  Raw gRPC Benchmark Summary (Go - WRITE)")
	fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
	fmt.Printf("  Total Batches:      %d\n", insertCount)
	fmt.Printf("  Batch Size:         %d\n", batchCount)
	fmt.Printf("  Total Records:      %d\n", totalRecords)
	fmt.Printf("  Total Inserts:      %d (includes %d duplicates)\n", totalInserts, alreadyExistsCount)
	fmt.Printf("  Concurrency:        %d\n", insertConcurrency)
	fmt.Printf("  Total Time:         %.2fms\n", float64(writeOverallDuration))
	fmt.Printf("  Inserts/Second:     %.2f\n", writeInsertsPerSecond)
	fmt.Printf("  Avg Batch Duration: %.2fms\n", writeAvg)
	fmt.Printf("  Min Batch Duration: %.2fms\n", writeMin)
	fmt.Printf("  Max Batch Duration: %.2fms\n", writeMax)
	fmt.Printf("  P50:                %.2fms\n", wp50)
	fmt.Printf("  P90:                %.2fms\n", wp90)
	fmt.Printf("  P95:                %.2fms\n", wp95)
	fmt.Printf("  P99:                %.2fms\n", wp99)
	fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n")
}

func protoValue(val interface{}) *structpb.Value {
	if val == nil {
		return &structpb.Value{Kind: &structpb.Value_NullValue{NullValue: structpb.NullValue_NULL_VALUE}}
	}
	switch v := val.(type) {
	case string:
		return &structpb.Value{Kind: &structpb.Value_StringValue{StringValue: v}}
	case *string:
		if v == nil {
			return &structpb.Value{Kind: &structpb.Value_NullValue{NullValue: structpb.NullValue_NULL_VALUE}}
		}
		return &structpb.Value{Kind: &structpb.Value_StringValue{StringValue: *v}}
	case int64:
		return &structpb.Value{Kind: &structpb.Value_StringValue{StringValue: strconv.FormatInt(v, 10)}}
	case *int64:
		if v == nil {
			return &structpb.Value{Kind: &structpb.Value_NullValue{NullValue: structpb.NullValue_NULL_VALUE}}
		}
		return &structpb.Value{Kind: &structpb.Value_StringValue{StringValue: strconv.FormatInt(*v, 10)}}
	default:
		return &structpb.Value{Kind: &structpb.Value_StringValue{StringValue: fmt.Sprintf("%v", v)}}
	}
}

func sha256Hash(str string) string {
	hash := sha256.Sum256([]byte(str))
	return hex.EncodeToString(hash[:])
}

func generateUUID() string {
	b := make([]byte, 16)
	rand.Read(b)
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
}

func percentile(sorted []time.Duration, fraction float64) time.Duration {
	if len(sorted) == 0 {
		return 0
	}
	index := int(float64(len(sorted)) * fraction)
	if index >= len(sorted) {
		index = len(sorted) - 1
	}
	return sorted[index]
}

func getEnv(key, fallback string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return fallback
}

func getEnvInt(key string, fallback int) int {
	if value := os.Getenv(key); value != "" {
		if intValue, err := strconv.Atoi(value); err == nil {
			return intValue
		}
	}
	return fallback
}

File: handwritten/spanner/issue/node/raw_grpc_benchmark.js
'use strict';

const { performance } = require('perf_hooks');
const crypto = require('crypto');
const grpc = require('@grpc/grpc-js');
const grpcGcp = require('grpc-gcp')(grpc);
const gcpApiConfig = require('@google-cloud/spanner/build/src/spanner_grpc_config.json');
const { v1 } = require('@google-cloud/spanner');

// Defaults match what is used in benchmark.js / benchmark-insert-repl.ts
const DB_PROJECT_ID = process.env.DB_PROJECT_ID || 'emulator';
const DB_INSTANCE = process.env.DB_INSTANCE || 'device-tracking';
const DB_DATABASE = process.env.DB_DATABASE || 'device-tracking';
const TABLE_NAME = process.env.TABLE_NAME || 'tracking.DeviceRecentActivityLog';
const READ_COUNT = envInt('READ_COUNT', 1000);
const CONCURRENCY = envInt('CONCURRENCY', 10);

// Write benchmark configuration
const SAMPLE_SIZE = envInt('SAMPLE_SIZE', 10000);
const INSERT_COUNT = envInt('INSERT_COUNT', 1000);
const INSERT_CONCURRENCY = envInt('INSERT_CONCURRENCY', 110);
const BATCH_COUNT = envInt('BATCH_COUNT', 1);
const DUPLICATE_INSERT = process.env.DUPLICATE_INSERT === 'true';

function sha256(str) {
  return crypto.createHash('sha256').update(str).digest('hex');
}

function protoValue(val) {
  if (val === null || val === undefined) {
    return { nullValue: 0 };
  }
  return { stringValue: val.toString() };
}

async function runBenchmark() {
  const databasePath = `projects/${DB_PROJECT_ID}/instances/${DB_INSTANCE}/databases/${DB_DATABASE}`;

  console.log(`Target Database: ${databasePath}`);
  console.log(`Target Table: ${TABLE_NAME}`);
  console.log(`Read Config: READ_COUNT=${READ_COUNT}, CONCURRENCY=${CONCURRENCY}`);
  console.log(`Write Config: SAMPLE_SIZE=${SAMPLE_SIZE}, INSERT_COUNT=${INSERT_COUNT}, INSERT_CONCURRENCY=${INSERT_CONCURRENCY}, BATCH_COUNT=${BATCH_COUNT}, DUPLICATE_INSERT=${DUPLICATE_INSERT}`);

  // Customize grpc-gcp channel pool configuration for a constant 4 channels
  const customGcpConfig = JSON.parse(JSON.stringify(gcpApiConfig));
  customGcpConfig.channelPool.minSize = 4;
  customGcpConfig.channelPool.maxSize = 4;
  customGcpConfig.channelPool.maxConcurrentStreamsLowWatermark = 1;

  // Initialize a single GAPIC client with grpc-gcp channel pooling enabled
  const client = new v1.SpannerClient({
    grpc,
    'grpc.callInvocationTransformer': grpcGcp.gcpCallInvocationTransformer,
    'grpc.channelFactoryOverride': grpcGcp.gcpChannelFactoryOverride,
    'grpc.gcpApiConfig': grpcGcp.createGcpApiConfig(customGcpConfig),
  });

  console.log('\nCreating 4 multiplexed sessions (one per channel) via raw gRPC...');
  const sessions = await Promise.all(
    Array.from({ length: 1 }, () =>
      client.createSession({
        database: databasePath,
        session: {
          multiplexed: true,
        },
      }).then(([s]) => s)
    )
  );
  console.log(`Created 4 Multiplexed Sessions.`);

  try {
    console.log('Fetching actual seeded keys from the database via SDK bootstrap to ensure 100% hit rate...');
    const { Spanner } = require('@google-cloud/spanner');
    const bootstrapSpanner = new Spanner({ projectId: DB_PROJECT_ID });
    const bootstrapDb = bootstrapSpanner.instance(DB_INSTANCE).database(DB_DATABASE);
    const [existingRows] = await bootstrapDb.run({
      sql: `SELECT deviceRecentActivityLogId FROM ${TABLE_NAME} LIMIT ${SAMPLE_SIZE}`
    });
    if (existingRows.length === 0) {
      throw new Error('No seeded rows found in the database! Please seed the database first.');
    }
    const targetKeys = existingRows.map(r => r.toJSON().deviceRecentActivityLogId);
    console.log(`Successfully retrieved ${targetKeys.length} active keys. Preparing read tasks...`);
    await bootstrapDb.close();
    bootstrapSpanner.close();

    // ==========================================
    // 1. READ BENCHMARK
    // ==========================================
    const readItems = Array.from({ length: READ_COUNT });
    const readDurations = new Array(READ_COUNT);

    console.log(`\nStarting READ benchmark (SQL ExecuteSql) with concurrency ${CONCURRENCY}...`);
    const readOverallStartTime = performance.now();

    await asyncMap(
      readItems,
      async (_, i) => {
        const startTime = performance.now();
        const session = sessions[i % sessions.length];
        const key = targetKeys[i % targetKeys.length];
        const request = {
          session: session.name,
          sql: `SELECT deviceRecentActivityLogId, deviceRecordId, deviceDetailsId, userId, createdAt FROM ${TABLE_NAME} WHERE deviceRecentActivityLogId = @id`,
          params: {
            fields: {
              id: {
                stringValue: key,
              },
            },
          },
          paramTypes: {
            id: {
              code: 6, // STRING
            },
          },
        };

        await client.executeSql(request);
        const duration = performance.now() - startTime;

        readDurations[i] = duration;
      },
      { concurrency: CONCURRENCY }
    );

    const readOverallDuration = performance.now() - readOverallStartTime;
    const readOpsPerSecond = READ_COUNT / (readOverallDuration / 1000);

    const readSum = readDurations.reduce((a, b) => a + b, 0);
    const readAvg = readSum / readDurations.length;

    const readSorted = [...readDurations].sort((a, b) => a - b);
    const rp50 = percentile(readSorted, 0.5);
    const rp90 = percentile(readSorted, 0.9);
    const rp95 = percentile(readSorted, 0.95);
    const rp99 = percentile(readSorted, 0.99);

    console.log('\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log('  Raw gRPC Benchmark Summary (Node.js - READ)');
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log(`  Total Reads:        ${READ_COUNT}`);
    console.log(`  Concurrency:        ${CONCURRENCY}`);
    console.log(`  Total Time:         ${readOverallDuration.toFixed(2)}ms`);
    console.log(`  Reads/Second:       ${readOpsPerSecond.toFixed(2)}`);
    console.log(`  Avg Latency:        ${readAvg.toFixed(2)}ms`);
    console.log(`  P50 Latency:        ${rp50.toFixed(2)}ms`);
    console.log(`  P90 Latency:        ${rp90.toFixed(2)}ms`);
    console.log(`  P95 Latency:        ${rp95.toFixed(2)}ms`);
    console.log(`  P99 Latency:        ${rp99.toFixed(2)}ms`);
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n');

    // ==========================================
    // 2. SLEEP 5 SECONDS
    // ==========================================
    console.log('Sleeping for 5 seconds before starting writes...');
    await new Promise(resolve => setTimeout(resolve, 5000));

    // ==========================================
    // 3. WRITE BENCHMARK
    // ==========================================
    console.log(`\nStarting WRITE benchmark with concurrency ${INSERT_CONCURRENCY}...`);

    const totalRecords = INSERT_COUNT * BATCH_COUNT;
    const totalInserts = DUPLICATE_INSERT ? totalRecords * 2 : totalRecords;

    // Pre-generate batches
    const batches = [];
    for (let i = 0; i < INSERT_COUNT; i++) {
      const batch = [];
      for (let j = 0; j < BATCH_COUNT; j++) {
        const idx = i * BATCH_COUNT + j;
        const deviceRecordId = sha256(`device-${idx % SAMPLE_SIZE}`);
        const deviceDetailsId = sha256(`detail-${idx % SAMPLE_SIZE}`);
        const httpRequestDetailsId = sha256(`request-${idx % SAMPLE_SIZE}`);
        const httpRequestLocationId = sha256(`location-${idx % SAMPLE_SIZE}`);
        const xRequestId = crypto.randomUUID();
        const sessionId = crypto.randomUUID();
        const createdAt = new Date(Date.now() + idx * 1000);

        const deviceRecentActivityLogId = sha256(`benchmark-${idx}${deviceDetailsId}${httpRequestDetailsId}${createdAt.getTime()}${sessionId}`);

        batch.push({
          deviceRecentActivityLogId,
          deviceRecordId,
          deviceDetailsId,
          httpRequestDetailsId,
          ipAddress: Buffer.from([192, 168, 1, 100]).toString('base64'),
          institutionId: 'benchmark-institution',
          userId: Math.random() > 0.5 ? 'user-123' : null,
          username: null,
          xRequestId,
          httpRequestLocationId,
          latency: Math.random() > 0.5 ? Math.floor(Math.random() * 100) : null,
          sessionId,
          createdAt: createdAt.toISOString(),
        });
      }
      batches.push(batch);
    }

    const writeDurations = [];
    let alreadyExistsCount = 0;
    const writeOverallStartTime = performance.now();

    await asyncMap(
      batches,
      async (batch, i) => {
        const startTime = performance.now();

        const mutations = batch.map(rec => ({
          insert: {
            table: TABLE_NAME,
            columns: [
              'deviceRecentActivityLogId',
              'deviceRecordId',
              'deviceDetailsId',
              'httpRequestDetailsId',
              'ipAddress',
              'institutionId',
              'userId',
              'username',
              'xRequestId',
              'httpRequestLocationId',
              'latency',
              'sessionId',
              'createdAt',
            ],
            values: [
              {
                values: [
                  protoValue(rec.deviceRecentActivityLogId),
                  protoValue(rec.deviceRecordId),
                  protoValue(rec.deviceDetailsId),
                  protoValue(rec.httpRequestDetailsId),
                  protoValue(rec.ipAddress),
                  protoValue(rec.institutionId),
                  protoValue(rec.userId),
                  protoValue(rec.username),
                  protoValue(rec.xRequestId),
                  protoValue(rec.httpRequestLocationId),
                  protoValue(rec.latency),
                  protoValue(rec.sessionId),
                  protoValue(rec.createdAt),
                ],
              },
            ],
          },
        }));

        const session = sessions[i % sessions.length];

        const commitRequest = {
          session: session.name,
          singleUseTransaction: {
            readWrite: {},
          },
          mutations,
        };

        try {
          await client.commit(commitRequest);
        } catch (e) {
          if (e.code === 6) { // ALREADY_EXISTS
            alreadyExistsCount++;
          } else {
            throw e;
          }
        }

        if (DUPLICATE_INSERT) {
          try {
            await client.commit(commitRequest);
          } catch (e) {
            if (e.code === 6) {
              alreadyExistsCount++;
            } else {
              throw e;
            }
          }
        }

        const duration = performance.now() - startTime;
        writeDurations.push(duration);

        if ((i + 1) % 100 === 0 || (i + 1) === INSERT_COUNT) {
          const slice = writeDurations.slice(-100);
          const avg = slice.reduce((a, b) => a + b, 0) / slice.length;
          console.log(`  [${i + 1}/${INSERT_COUNT}] Last batches avg: ${avg.toFixed(2)}ms (${BATCH_COUNT} records/batch)`);
        }
      },
      { concurrency: INSERT_CONCURRENCY }
    );

    const writeOverallDuration = performance.now() - writeOverallStartTime;
    const writeInsertsPerSecond = totalInserts / (writeOverallDuration / 1000);

    const writeSum = writeDurations.reduce((a, b) => a + b, 0);
    const writeAvg = writeSum / writeDurations.length;
    const writeMin = Math.min(...writeDurations);
    const writeMax = Math.max(...writeDurations);

    const writeSorted = [...writeDurations].sort((a, b) => a - b);
    const wp50 = percentile(writeSorted, 0.5);
    const wp90 = percentile(writeSorted, 0.9);
    const wp95 = percentile(writeSorted, 0.95);
    const wp99 = percentile(writeSorted, 0.99);

    console.log('\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log('  Raw gRPC Benchmark Summary (Node.js - WRITE)');
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log(`  Total Batches:      ${INSERT_COUNT}`);
    console.log(`  Batch Size:         ${BATCH_COUNT}`);
    console.log(`  Total Records:      ${totalRecords}`);
    console.log(`  Total Inserts:      ${totalInserts} (includes ${alreadyExistsCount} duplicates)`);
    console.log(`  Concurrency:        ${INSERT_CONCURRENCY}`);
    console.log(`  Total Time:         ${writeOverallDuration.toFixed(2)}ms`);
    console.log(`  Inserts/Second:     ${writeInsertsPerSecond.toFixed(2)}`);
    console.log(`  Avg Batch Duration: ${writeAvg.toFixed(2)}ms`);
    console.log(`  Min Batch Duration: ${writeMin.toFixed(2)}ms`);
    console.log(`  Max Batch Duration: ${writeMax.toFixed(2)}ms`);
    console.log(`  P50:                ${wp50.toFixed(2)}ms`);
    console.log(`  P90:                ${wp90.toFixed(2)}ms`);
    console.log(`  P95:                ${wp95.toFixed(2)}ms`);
    console.log(`  P99:                ${wp99.toFixed(2)}ms`);
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n');

  } finally {
    console.log('\nBenchmark finished. Multiplexed session is left to be managed by server.');
  }
}

async function asyncMap(items, mapperFn, opts = {}) {
  const concurrency = Math.max(1, opts.concurrency ?? Infinity);
  const results = new Array(items.length);
  let currentIndex = 0;

  async function processNext() {
    while (currentIndex < items.length) {
      const index = currentIndex++;
      results[index] = await mapperFn(items[index], index);
    }
  }

  const workers = Array.from({ length: Math.min(concurrency, items.length) }, () => processNext());
  await Promise.all(workers);
  return results;
}

function percentile(sorted, fraction) {
  if (!sorted.length) return 0;
  const index = Math.min(sorted.length - 1, Math.floor(sorted.length * fraction));
  return sorted[index];
}

function envInt(key, fallback) {
  const value = process.env[key];
  return value === undefined || value === '' ? fallback : parseInt(value, 10);
}

runBenchmark().catch(console.error);
